src/Init/Data/Json/Parser.lean

```diff
@@ -0,0 +1,332 @@
+/-
+Copyright (c) 2019 Gabriel Ebner. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Authors: Gabriel Ebner, Marc Huisinga
+-/
+prelude
+import Init.Data.Json.Basic 
+import Init.Control.Except
+
+inductive Quickparse.Result (α : Type)
+| success {} (pos : String.Iterator) (res : α) : Quickparse.Result
+| error {} (pos : String.Iterator) (err : String) : Quickparse.Result
+
+def Quickparse (α : Type) : Type := String.Iterator → Quickparse.Result α
+
+def String.Iterator.skipWsCore : Nat → String.Iterator → String.Iterator
+| 0  , it => it
+| i+1, it =>
+  let c := it.curr;
+  if c = '\u0009' ∨ c = '\u000a' ∨ c = '\u000d' ∨ c = '\u0020' then
+    it.next.skipWsCore i
+  else
+    it
+
+def String.Iterator.skipWs (it : String.Iterator) : String.Iterator :=
+it.skipWsCore (it.s.length - it.i)
+
+instance Quickparse.inhabited (α : Type) : Inhabited (Quickparse α) := 
+⟨fun it => Quickparse.Result.error it ""⟩
+
+namespace Quickparse
+
+open Result
+
+@[inline]
+protected def pure {α : Type} (a : α) : Quickparse α | it => 
+success it a
+
+@[inline]
+protected def bind {α β : Type} (f : Quickparse α) (g : α → Quickparse β) : Quickparse β | it => 
+match f it with
+| success rem a => g a rem
+| error pos msg => error pos msg
+
+@[inline]
+def fail {α : Type} (msg : String) : Quickparse α | it => 
+error it msg
+
+@[inline]
+instance : Monad Quickparse :=
+{ pure := @Quickparse.pure, bind := @Quickparse.bind }
+
+@[inline]
+def remaining : Quickparse Nat | it => 
+success it (it.s.length - it.i)
+
+def unexpectedEndOfInput := "unexpected end of input"
+
+@[inline]
+def peek? : Quickparse (Option Char) | it => 
+if it.hasNext then 
+  success it it.curr 
+else 
+  success it none
+
+@[inline]
+def peek! : Quickparse Char := do
+some c <- peek? | fail unexpectedEndOfInput; 
+pure c 
+
+@[inline]
+def skip : Quickparse Unit | it => 
+success it.next ()
+
+@[inline]
+def next : Quickparse Char := do
+c ← peek!; skip;
+pure c
+
+def expect (s : String) : Quickparse Unit | it =>
+if it.extract (it.forward s.length) = s then
+  success (it.forward s.length) ()
+else
+  error it ("expected: " ++ s)
+
+@[inline]
+def ws : Quickparse Unit | it => 
+success it.skipWs ()
+
+def expectedEndOfInput := "expected end of input"
+
+@[inline]
+def eoi : Quickparse Unit | it =>
+if it.hasNext then
+  error it expectedEndOfInput
+else
+  success it ()
+
+end Quickparse
+
+namespace Json.Parser
+
+open Quickparse
+
+def invalidHexCh : String := "invalid hex character"
+
+@[inline]
+def hexChar : Quickparse Nat := do
+c ← next;
+if '0' ≤ c ∧ c ≤ '9' then
+  pure $ c.val.toNat - '0'.val.toNat
+else if 'a' ≤ c ∧ c ≤ 'f' then
+  pure $ c.val.toNat - 'a'.val.toNat
+else if 'A' ≤ c ∧ c ≤ 'F' then
+  pure $ c.val.toNat - 'A'.val.toNat
+else
+  fail invalidHexCh
+
+@[inline]
```

Comment:
I would move the `if` and its hot path (`c1 /= '\\') into `strCore` and then remove the `[inline]`